File: java/year2021/month1/no39/CombinationSum.java
package year2021.month1.no39;

import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

public class CombinationSum {
    public static void main(String[] args) {
        int[] candidates1 = new int[]{2, 3, 6, 7};
        int target1 = 7;
        int[] candidates2 = new int[]{2, 3, 5};
        int target2 = 8;
        int[] candidates3 = new int[]{2, 7, 6, 3, 5, 1};
        int target3 = 9;
        System.out.println(combinationSum1(candidates1, target1));
        System.out.println(combinationSum1(candidates2, target2));
        System.out.println(combinationSum1(candidates3, target3));
    }

    public static List<List<Integer>> combinationSum1(int[] candidates, int target) {
        List<List<Integer>> result = new ArrayList<>();
        if (candidates == null || candidates.length == 0) {
            return result;
        }
        backtrack1(candidates, 0, target, result, new LinkedList<>());
        return result;
    }

    private static void backtrack1(int[] candidates, int pos, int target, List<List<Integer>> result, Deque<Integer> list) {
        if (target == 0) {
            result.add(new ArrayList<>(list));
            return;
        }
        if (pos >= candidates.length) {
            return;
        }
        int loop = target / candidates[pos];
        for (int i = 0; i <= loop; i++) {
            int num = i * candidates[pos];
            for (int j = 0; j < i; j++) {
                list.offerLast(candidates[pos]);
            }
            backtrack1(candidates, pos + 1, target - num, result, list);
            for (int j = 0; j < i; j++) {
                list.pollLast();
            }
        }
    }

    public static List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> result = new ArrayList<>();
        if (candidates == null || candidates.length == 0) {
            return result;
        }
        Arrays.sort(candidates);
        backtrack1(candidates, 0, target, result, new LinkedList<>());
        return result;
    }

    private static void backtrack(int[] candidates, int pos, int target, List<List<Integer>> result, int[] array, int index) {
        if (target == 0) {
            result.add(IntStream.of(array).boxed().filter(num -> num > 0).collect(Collectors.toList()));
            return;
        }
        if (pos >= candidates.length) {
            return;
        }
        if (candidates[pos] > target) {
            return;
        }
        int loop = target / candidates[pos];
        for (int i = 0; i <= loop; i++) {
            int num = i * candidates[pos];
            for (int j = 0; j < i; j++) {
                array[index++] = candidates[pos];
            }
            backtrack(candidates, pos + 1, target - num, result, array, index);
            for (int j = 0; j < i; j++) {
                array[--index] = 0;
            }
        }
    }
}

/*
* 给定一个无重复元素的数组 candidates 和一个目标数 target ，找出 candidates 中所有可以使数字和为 target 的组合。

candidates 中的数字可以无限制重复被选取。

说明：

所有数字（包括 target）都是正整数。
解集不能包含重复的组合。 
示例 1：

输入：candidates = [2,3,6,7], target = 7,
所求解集为：
[
  [7],
  [2,2,3]
]
示例 2：

输入：candidates = [2,3,5], target = 8,
所求解集为：
[
  [2,2,2,2],
  [2,3,3],
  [3,5]
]
 

提示：

1 <= candidates.length <= 30
1 <= candidates[i] <= 200
candidate 中的每个元素都是独一无二的。
1 <= target <= 500

来源：力扣（LeetCode）
链接：https://leetcode-cn.com/problems/combination-sum
著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。
* */

File: go/year2025/month4/no39/CombinationSum.go
package main

import (
	"fmt"
	"sort"
)

func main() {
	candidates1 := []int{2, 3, 6, 7}
	target1 := 7
	candidates2 := []int{2, 3, 5}
	target2 := 8
	candidates3 := []int{2}
	target3 := 1
	println(fmt.Sprintf("%+v", combinationSum(candidates1, target1)))
	println(fmt.Sprintf("%+v", combinationSum(candidates2, target2)))
	println(fmt.Sprintf("%+v", combinationSum(candidates3, target3)))
}

func combinationSum(candidates []int, target int) [][]int {
	ans := make([][]int, 0)
	arr := make([]int, 0)
	backtrack(&ans, candidates, target, &arr, 0)
	return ans
}

// backtrack 通过 pos 控制 candidates 中每个元素的选取次数，避免重复答案
func backtrack(ans *[][]int, candidates []int, target int, arr *[]int, pos int) {
	if target == 0 {
		if len(*arr) > 0 {
			p := make([]int, 0, len(*arr))
			for _, val := range *arr {
				p = append(p, val)
			}
			*ans = append(*ans, p)
		}
		return
	}
	if pos >= len(candidates) {
		return
	}
	loop := target / candidates[pos]
	for i := 0; i <= loop; i++ {
		for j := 0; j < i; j++ { // 选 i 个 candidates[pos]
			*arr = append(*arr, candidates[pos])
		}
		sum := i * candidates[pos]
		backtrack(ans, candidates, target-sum, arr, pos+1)
		*arr = (*arr)[:len(*arr)-i] // 回溯
	}
}

func combinationSum1(candidates []int, target int) [][]int {
	ans := make([][]int, 0)
	arr := make([]int, 0)
	// backtrack
	backtrack1(&ans, candidates, target, &arr, 0)
	// deduplicate
	return deduplicate(ans)
}

func deduplicate(arr [][]int) (res [][]int) {
	dic := make(map[string][]int)
	for _, s := range arr {
		sort.Ints(s)
		key := fmt.Sprintf("%+v", s)
		dic[key] = s
	}
	for _, val := range dic {
		res = append(res, val)
	}
	return
}

func backtrack1(ans *[][]int, candidates []int, target int, arr *[]int, sum int) {
	if sum > target {
		// candidates 都是正数，sum 超过时，后续不可能满足条件
		return
	}
	if sum == target {
		if len(*arr) > 0 {
			p := make([]int, 0, len(*arr))
			for _, val := range *arr {
				p = append(p, val)
			}
			*ans = append(*ans, p)
		}
	}
	for _, candidate := range candidates {
		if sum+candidate > target {
			continue
		}
		*arr = append(*arr, candidate)
		backtrack1(ans, candidates, target, arr, sum+candidate)
		*arr = (*arr)[:len(*arr)-1]
	}
}

/*
 给你一个 无重复元素 的整数数组 candidates 和一个目标整数 target ，
找出 candidates 中可以使数字和为目标数 target 的 所有 不同组合 ，
并以列表形式返回。你可以按 任意顺序 返回这些组合。

candidates 中的 同一个 数字可以 无限制重复被选取 。如果至少一个数字的被选数量不同，则两种组合是不同的。
对于给定的输入，保证和为 target 的不同组合数少于 150 个。

示例 1：
输入：candidates = [2,3,6,7], target = 7
输出：[[2,2,3],[7]]
解释：
2 和 3 可以形成一组候选，2 + 2 + 3 = 7 。注意 2 可以使用多次。
7 也是一个候选， 7 = 7 。
仅有这两种组合。

示例 2：
输入: candidates = [2,3,5], target = 8
输出: [[2,2,2,2],[2,3,3],[3,5]]

示例 3：
输入: candidates = [2], target = 1
输出: []


提示：
1 <= candidates.length <= 30
2 <= candidates[i] <= 40
candidates 的所有元素 互不相同
1 <= target <= 40
*/
